test/performance/odp_dmafwd.c

```diff
@@ -1190,6 +1190,12 @@ static int process_packets(void *args)
 				pkt_fn(pkt_vec->pkts, burst_size, pkt_vec->pktio, init_fn,
 				       start_fn, config);
 				pop_packets(pkt_vec, burst_size);
+			} else {
+				if (pkt_vec->num) {
+					pkt_fn(pkt_vec->pkts, pkt_vec->num, pkt_vec->pktio, init_fn,
+					       start_fn, config);
+					pop_packets(pkt_vec, pkt_vec->num);
+				}
```

Comment:
The gathering-to-burst is by design in order to test the characteristics of DMA vs. SW copy with different size bursts so the burst has to be formed first before sending anything (which might take a few scheduling rounds).

Was this causing some issues?

Reply:
If the number of packets to be transmitted are not in multiples of the burst size then the remaining packets were not transmitted. For example: Number of packets to be sent out is 100 and burst size is 15 then 10 packets will not be transmitted.

Reply:
Yes, that is conceptually fine. The remaining packets should be handled in `free_pending_packets()` which frees these unsent packets when test is done.

Are you seeing some issues or leaking resources without this change?

Reply:
Instead of freeing the remaining packets in `free_pending_packets()`, one option could be to transmit them, so no packets would be dropped from the point of view of an outside observer.

Reply:
Well, there's really no need to send the packets just for that. We have several testers that do a "drain-and-free" phase at the end, after the measured test, which might result in fewer received packets than sent.

Reply:
From the test statistics it is confusing as in the stats it neither shows dropped nor forwarded. In the below statistics snippet 100 packets were sent with burst size of 15:

`DMA forwarder done

    copy mode:       SW
    burst size:      15
    packet length:   1024
    max cache size:  8

    worker 0:
        packet copy errors: 0
        packets forwarded:  90
        packets dropped:    0
        call cycles per schedule round:
            total:    3999
            schedule: 3959
            rounds:   9071500`

Reply:
As a compromise, `free_pending_packets()` could add the drops to the `stats_t::discards`.

Reply:
That will show up in stats as packets dropped and will be a wrong interpretation for the packets not sent.

Reply:
In any case, the current proposal breaks the core bursting logic, so it can't be part of the main test loop.

Reply:
If using `stats_t::discards` counter is too ambiguous, another option could be to add a new counter for clean up. E.g. `packets cleaned up:`

Btw. Tuomas is out of office this week.